libr/anal/block.c

```diff
@@ -0,0 +1,579 @@
+	while (node && addr < node->addr) {
+		// less than the current node, but might still be contained further down
+		node = unwrap (node->_rb.child[0]);
+	}
+	if (!node) {
+		return true;
+	}
+	if (addr >= node->_max_end) {
+		return true;
+	}
+	if (addr < node->addr + node->size) {
+		if (!cb (node, user)) {
+			return false;
+		}
+	}
+	// This can be done more efficiently by building the stack manually
+	if (!all_in (unwrap (node->_rb.child[0]), addr, cb, user)) {
+		return false;
+	}
+	if (!all_in (unwrap (node->_rb.child[1]), addr, cb, user)) {
+		return false;
+	}
+	return true;
+}
+
+R_API bool r_anal_block_get_in(RAnal *anal, ut64 addr, RAnalBlockCb cb, void *user) {
+	return all_in (anal->bb_tree ? unwrap (anal->bb_tree) : NULL, addr, cb, user);
+}
+
+static bool block_list_cb(RAnalBlock *block, void *user) {
+	RList *list = user;
+	r_anal_block_ref (block);
+	r_list_push (list, block);
+	return true;
+}
+
+R_API RList *r_anal_block_get_in_list(RAnal *anal, ut64 addr) {
+	RList *list = r_list_newf ((RListFree)r_anal_block_unref);
+	if (!list) {
+		return NULL;
+	}
+	r_anal_block_get_in (anal, addr, block_list_cb, list);
+	return list;
+}
+
+static void all_intersect(RAnalBlock *node, ut64 addr, ut64 size, RAnalBlockCb cb, void *user) {
+	ut64 end = addr + size;
+	while (node && end <= node->addr) {
+		// less than the current node, but might still be contained further down
+		node = unwrap (node->_rb.child[0]);
+	}
+	if (!node) {
+		return;
+	}
+	if (addr >= node->_max_end) {
+		return;
+	}
+	if (addr < node->addr + node->size) {
+		cb (node, user);
+	}
+	// This can be done more efficiently by building the stack manually
+	all_intersect (unwrap (node->_rb.child[0]), addr, size, cb, user);
+	all_intersect (unwrap (node->_rb.child[1]), addr, size, cb, user);
+}
+
+R_API void r_anal_blocks_get_intersect(RAnal *anal, ut64 addr, ut64 size, RAnalBlockCb cb, void *user) {
+	all_intersect (anal->bb_tree ? unwrap (anal->bb_tree) : NULL, addr, size, cb, user);
+}
+
+R_API RList *r_anal_block_get_intersect_list(RAnal *anal, ut64 addr, ut64 size) {
+	RList *list = r_list_newf ((RListFree)r_anal_block_unref);
+	if (!list) {
+		return NULL;
+	}
+	r_anal_blocks_get_intersect (anal, addr, size, block_list_cb, list);
+	return list;
+}
+
+R_API RAnalBlock *r_anal_block_create(RAnal *anal, ut64 addr, ut64 size) {
+	if (r_anal_block_get_at (anal, addr)) {
+		return NULL;
+	}
+	RAnalBlock *block = block_new (anal, addr, size);
+	if (!block) {
+		return NULL;
+	}
+	r_rbtree_aug_insert (&anal->bb_tree, &block->addr, &block->_rb, __bb_addr_cmp, NULL, __max_end);
+	return block;
+}
+
+R_API void r_anal_block_delete(RAnalBlock *bb) {
+	RAnal *anal = bb->anal;
+	r_anal_block_ref (bb);
+	while (!r_list_empty (bb->fcns)) {
+		r_anal_function_block_remove (r_list_first (bb->fcns), bb);
+	}
+	r_anal_block_unref (bb);
+	r_anal_block_check_invariants (anal);
+}
+
+R_API void r_anal_block_set_size(RAnalBlock *block, ut64 size) {
```

Comment:
If we set a new size of the RAnalBlock we should probably also update `op_pos`, no?

Reply:
Depends. This is more a low-level call that only changes the size of the block while filling `op_pos` is more of an analysis.

Reply:
If it is so, I don't expect it to be a public R_API. I think it is very easy to forget updating op_pos (or other stuff that depend on the size of the bb).

Reply:
No, there must be an interface like that for the kind of analysis that `fcn_recurse()` does. Otherwise multiple r_anal_op passes would be necessary to do on the same op.
A good way to avoid the possibility of forgetting to update it would be to make it lazy, but that is outside the scope of this pr.